src/Dialect/ONNX/Transforms/FusePadAvgpool.cpp

```diff
@@ -0,0 +1,141 @@
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Dialect/ONNX/ONNXOps/OpHelper.hpp"
+#include "src/Pass/Passes.hpp"
+
+using namespace mlir;
+using namespace onnx_mlir;
+
+namespace {
+
+struct FusePadIntoAveragePoolPattern
+    : public OpRewritePattern<ONNXAveragePoolOp> {
+  using OpRewritePattern<ONNXAveragePoolOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      ONNXAveragePoolOp avgOp, PatternRewriter &rewriter) const override {
+
+    Value input = avgOp.getX();
+    auto padOp = input.getDefiningOp<ONNXPadOp>();
+    if (!padOp)
+      return failure();
+
+    StringAttr modeAttr = padOp.getModeAttr();
+    StringRef mode = "constant";
+    if (modeAttr)
+      mode = modeAttr.getValue();
+    if (mode != "constant")
+      return failure();
+    float padValue = 0.0f;
+
+    Value padsInput = padOp.getPads();
+    Value constantValInput = padOp.getConstantValue();
+
+    auto padsConstOp =
+        dyn_cast_or_null<ONNXConstantOp>(padsInput.getDefiningOp());
+    if (!padsConstOp)
+      return failure();
+    auto padsAttr = dyn_cast_or_null<ElementsAttr>(padsConstOp.getValueAttr());
+    if (!padsAttr)
+      return failure();
+
+    auto constOp =
+        dyn_cast_or_null<ONNXConstantOp>(constantValInput.getDefiningOp());
+    if (!constOp)
+      return failure();
+    auto constAttr = dyn_cast_or_null<ElementsAttr>(constOp.getValueAttr());
+
+    if (!constAttr)
+      return failure();
+
+    auto firstAttr = *constAttr.getValues<Attribute>().begin();
+    if (auto fAttr = mlir::dyn_cast<FloatAttr>(firstAttr))
+      padValue = fAttr.getValueAsDouble();
+
+    if (padValue != 0.0f)
+      return failure();
+
+    SmallVector<int64_t> padsVals;
+    for (auto val : padsAttr.getValues<Attribute>()) {
+      if (auto iAttr = mlir::dyn_cast<IntegerAttr>(val)) {
+        auto pad = iAttr.getInt();
+        padsVals.push_back(pad);
+      } else {
+        padsVals.push_back(0);
+      }
```

Comment:
Isnt't it guaranteed by the spec that the padsVals are integers? 
```suggestion
        auto pad = cast<IntegerAttr>(val).getInt();
        padsVals.push_back(pad);

```

Reply:
Done